ga4gh/backend.py

```diff
@@ -55,71 +55,99 @@ def _getVariantSet(request, variantSetIdMap):
     """
     Implements generator logic for types which accept a start/end
-    range to search for the object
+    range to search for the object. Returns an iterator over
+    (object, pageToken) pairs. The pageToken is a string which allows
+    us to pick up the iteration at any point, and is None for the last
+    value in the iterator.
     """
     def __init__(self, request, containerIdMap):
         self._request = request
         self._containerIdMap = containerIdMap
-        self._badPageTokenExceptionMessage = (
-            "Inconsistent page token provided")
         self._container = self._getContainer()
-        self._startPosition, self._equalPositionsToSkip = \
-            self._getIntervalCounters()
-        self._iterator = self._getIterator()
-        self._generator = self._internalIterator()
-
-    def __iter__(self):
-        return self._generator
+        self._searchIterator = None
+        self._currentObject = None
+        self._nextObject = None
+        self._searchAnchor = None
+        self._distanceFromAnchor = None
+        if request.pageToken is None:
+            self._initialiseIteration()
+        else:
+            # Set the search start point and the number of records to skip from
+            # the page token.
+            searchAnchor, objectsToSkip = _parsePageToken(request.pageToken, 2)
+            self._pickUpIteration(searchAnchor, objectsToSkip)
+
+    def _initialiseIteration(self):
+        """
+        Starts a new iteration.
+        """
+        self._searchIterator = self._search(
+                self._request.start, self._request.end)
+        self._currentObject = next(self._searchIterator, None)
+        if self._currentObject is not None:
+            self._nextObject = next(self._searchIterator, None)
+            self._searchAnchor = self._request.start
+            self._distanceFromAnchor = 0
+            firstObjectStart = self._getStart(self._currentObject)
+            if firstObjectStart > self._request.start:
+                self._searchAnchor = firstObjectStart
+
+    def _pickUpIteration(self, searchAnchor, objectsToSkip):
+        """
+        Picks up iteration from a previously provided page token. There are two
+        different phases here:
+        1) We are iterating over the initial set of intervals in which start
+        is < the search start coorindate.
+        2) We are iterating over the remaining intervals in which start >= to
+        the search start coordinate.
+        """
+        self._searchAnchor = searchAnchor
+        self._distanceFromAnchor = objectsToSkip
+        self._searchIterator = self._search(searchAnchor, self._request.end)
+        obj = next(self._searchIterator)
+        if searchAnchor == self._request.start:
+            # This is the initial set of intervals, we just skip forward
+            # objectsToSkip positions
+            for _ in range(objectsToSkip):
+                obj = next(self._searchIterator)
+        else:
```

Comment:
No, because we set searchAnchor = request.start for the first set of intervals (i.e., those with start <= request.start), and we don't update it until start becomes > request.start. This is the signalling condition to let us know we are in the initial phase.